Add synchronous pull PubSubMessageSource (#1321)
spring-cloud/spring-cloud-gcp: docs/src/main/asciidoc/spring-integration-pubsub.adoc

}
----

==== Pollable Message Source (using Pub/Sub Synchronous Pull)

While `PubSubInboundChannelAdapter`, through the underlying Asynchronous Pull Pub/Sub mechanism, provides the best performance for high-volume applications that receive a steady flow of messages, it can create load balancing anomalies due to message caching.
This behavior is most obvious when publishing a large batch of small messages that take a long time to process individually. It manifests as one subscriber taking up most messages, even if multiple subscribers are available to take on the work.
For a more detailed explanation of this scenario, see https://cloud.google.com/pubsub/docs/pull#dealing-with-large-backlogs-of-small-messages[GCP Pub/Sub documentation].

> In such a scenario, `PubSubMessageSource`, combined with a pollable channel, help spread the load between different subscribers more evenly.

Reviewer: Well, actually a `MessageSource` can be configured with an XML, as well as Java DSL.
It is not going to be fair to assume that `PubSubMessageSource` is going to work only with the `@InboundChannelAdapter`.

I would say it more generic way:
```suggestion
In such a scenario, a `PubSubMessageSource` can help spread the load between different subscribers more evenly.
```
And it is already doesn't matter for us how it is configured or even used directly, e.g. from the `MessageSourcePollingTemplate`.

}
----

==== Pollable Message Source (using Pub/Sub Synchronous Pull)

While `PubSubInboundChannelAdapter`, through the underlying Asynchronous Pull Pub/Sub mechanism, provides the best performance for high-volume applications that receive a steady flow of messages, it can create load balancing anomalies due to message caching.
This behavior is most obvious when publishing a large batch of small messages that take a long time to process individually.
It manifests as one subscriber taking up most messages, even if multiple subscribers are available to take on the work.
For a more detailed explanation of this scenario, see https://cloud.google.com/pubsub/docs/pull#dealing-with-large-backlogs-of-small-messages[GCP Pub/Sub documentation].

In such a scenario, a `PubSubMessageSource` can help spread the load between different subscribers more evenly.

As with the Inbound Channel Adapter, the message source has a configurable acknowledgement mode, payload type, and header mapping.
The default behavior is to return from the synchronous pull operation immediately if no messages are present.

> This can be overridden by using `blockOnPull()` method to wait for at least one message to arrive.

Author: @meltsufin Perhaps we should not be exposing this option at all. MessageSource is used by a polling channel -- what is the usecase in which a single poll needs to block? If nothing is returned, that's fine -- the channel will poll again shortly.

> This can be overridden by using `blockOnPull()` method to wait for at least one message to arrive.

Author: Tested this; poller behaves sensibly and waits for the previous poll to complete, even with `fixedRate`. Leaving block on pull option in the code.

[source,java]
----
@Bean
@InboundChannelAdapter(channel = "pubsubInputChannel", poller = @Poller(fixedDelay = "100"))
public MessageSource<Object> pubsubAdapter(PubSubTemplate pubSubTemplate) {
	PubSubMessageSource messageSource = new PubSubMessageSource(pubSubTemplate,  "exampleSubscription");
	messageSource.setAckMode(AckMode.MANUAL);
	messageSource.setPayloadType(String.class);

> 	messageSource.blockOnPull();

Reviewer 2: What about an example an discussion of fetch size?

The default behavior is to return from the synchronous pull operation immediately if no messages are present.
This can be overridden by using `setBlockOnPull()` method to wait for at least one message to arrive.

By default, `PubSubMessageSource` pulls from the subscription one message at a time.
To pull a batch of messages on each request, use the `setMaxFetchSize()` method to set the batch size.

[source,java]
----
@Bean
@InboundChannelAdapter(channel = "pubsubInputChannel", poller = @Poller(fixedDelay = "100"))
public MessageSource<Object> pubsubAdapter(PubSubTemplate pubSubTemplate) {
	PubSubMessageSource messageSource = new PubSubMessageSource(pubSubTemplate,  "exampleSubscription");
	messageSource.setAckMode(AckMode.MANUAL);
	messageSource.setPayloadType(String.class);
	messageSource.setBlockOnPull(true);
	messageSource.setMaxFetchSize(100);
	return messageSource;
}
----

The `MessageSource` attaches a `AcknowledgeablePubSubMessage` object to the `Message` headers.
Users can extract the `AcknowledgeablePubSubMessage` using the `GcpPubSubHeaders.ORIGINAL_MESSAGE` key and use it to (n)ack a message.

[source,java]
----

> @ServiceActivator(inputChannel = "pubsubInputChannel")

Author: The async example has it; but it's far below the adapter configuration -- after the ack mode discussion.

public void messageReceiver(String payload,
        @Header(GcpPubSubHeaders.ORIGINAL_MESSAGE) AcknowledgeablePubsubMessage message)
            throws InterruptedException {
    LOGGER.info("Message arrived by Synchronous Pull! Payload: " + payload);
    message.ack();
}
----

> NOTE: `AcknowledgeablePubSubMessage` objects acquired by synchronous pull are aware of their own acknowledgement IDs. Streaming pull does not expose this information due to limitations of the underlying API, and returns `BasicAcknowledgeablePubsubMessage` objects that allow acking/nacking individual messages, but not extracking acknowledgement IDs for future processing.

Reviewer: One sentence per line.
Also there is a type in the `extracking` word 😄
